localstack-core/localstack/services/events/event_rule_engine.py

```diff
@@ -0,0 +1,566 @@
+import ipaddress
+import json
+import re
+import typing as t
+
+from localstack.aws.api.events import InvalidEventPatternException
+
+
+class EventRuleEngine:
+    def evaluate_pattern_on_event(self, event_pattern: dict, message_body: str | dict):
```

Comment:
How does it work in ESM and Pipes if the pattern is invalid? I didn't really find it. 
The issue is that if the pattern is invalid, the whole matching logic will fail down the line, and it's part of the "compilation": it does not only validate, but compile the rules, unfolding the `$or` and such. I'll try to document that better

Reply:
I've renamed the `Validator` into `Compiler` but it's closer to what it really does, it translates the rules into something the engine can properly interprets. I guess that for ESM/Pipes, those rules would need to be evaluated only once, and could be cached then, instead of calling `matches_event` directly that way. I'm sure we could have some improvements, we should look at it in a follow up!